### PR TITLE
p2p: Refactor, move checksum verification into net, without changing behavior

### PR DESCRIPTION
**Pull request by troygiorshev**
This is an alternative approach to #15206 , without changing the disconnect behavior. I agree with this [comment](https://github.com/bitcoin/bitcoin/pull/15206#discussion_r428042604), and this PR is the other alternative, keeping the changes in net.cpp.

### CODE REVIEW

File: src/net.cpp [hunk 1]
```diff
@@ -686,7 +689,7 @@ const uint256& V1TransportDeserializer::GetMessageHash() const
     return data_hash;
 }
 
-CNetMessage V1TransportDeserializer::GetMessage(const CMessageHeader::MessageStartChars& message_start, int64_t time) {
+boost::optional<CNetMessage> V1TransportDeserializer::GetMessage(const CMessageHeader::MessageStartChars& message_start, int64_t time) {
```
Comment (dongcarl) [inline src/net.cpp:702]: We can use our own `Optional` class and its corresponding `nullopt` from `src/optional.h` :relaxed:

File: src/net_processing.cpp [hunk 2]
```diff
@@ -3602,20 +3602,11 @@ bool PeerLogicValidation::ProcessMessages(CNode* pfrom, std::atomic<bool>& inter
...
     unsigned int nMessageSize = msg.m_message_size;
 
-    // Checksum
-    CDataStream& vRecv = msg.m_recv;
-    if (!msg.m_valid_checksum)
-    {
-        LogPrint(BCLog::NET, "%s(%s, %u bytes): CHECKSUM ERROR peer=%d\n", __func__,
-           SanitizeString(msg_type), nMessageSize, pfrom->GetId());
-        return fMoreWork;
-    }
```
Comment (dongcarl) [inline src/net_processing.cpp:3612]: Nit: This, along with the removal of `m_valid_checksum` can probably be its own commit

File: src/net.cpp [hunk 3]
```diff
@@ -703,12 +706,14 @@ CNetMessage V1TransportDeserializer::GetMessage(const CMessageHeader::MessageSta
...
 
-    msg.m_valid_checksum = (memcmp(hash.begin(), hdr.pchChecksum, CMessageHeader::CHECKSUM_SIZE) == 0);
-    if (!msg.m_valid_checksum) {
+    bool valid_checksum = (memcmp(hash.begin(), hdr.pchChecksum, CMessageHeader::CHECKSUM_SIZE) == 0);
+    if (!valid_checksum) {
         LogPrint(BCLog::NET, "CHECKSUM ERROR (%s, %u bytes), expected %s was %s\n",
                  SanitizeString(msg.m_command), msg.m_message_size,
                  HexStr(hash.begin(), hash.begin()+CMessageHeader::CHECKSUM_SIZE),
                  HexStr(hdr.pchChecksum, hdr.pchChecksum+CMessageHeader::CHECKSUM_SIZE));
+        Reset();
```
Comment (dongcarl) [inline src/net.cpp:725]: Does this change behaviour in any way?

**Review by dongcarl — COMMENTED** [review]
Glad to see we can do this simplification without changing behaviour! Seems more correct.

File: src/net.cpp (hunk 3 above)
Comment (troygiorshev) [reply src/net.cpp:725]: Without this we fall into an infinite loop in one of the tests.  It ultimately stems from a combo of a few different members of V1TransportDeserializer interacting in such a way that we repeatedly fill and drain vRecv by the same amounts, without causing it to grow infinitely and without writing off of the end.

**Comment by troygiorshev** [comment]
Closed in favor of #3, which is now [bitcoin#19107](https://github.com/bitcoin/bitcoin/pull/19107)
